core/src/scene/filters.cpp

```diff
@@ -1,11 +1,214 @@
+               f.global != FilterGlobal::undefined,
+               f.key.c_str(), f.min, f.max);
+        return;
+    }
+    case Data::type<Function>::value: {
+        logMsg("%*s function\n", _indent, "");
+        break;
+    }
+    default:
+        break;
+    }
+
+}
+
+
+int Filter::matchCost() const {
+    // Add some extra penalty for set vs simple filters
+    int sum = -100;
+
+    switch (data.get_type_index()) {
+    case Data::type<OperatorAny>::value:
+        for (auto& f : operands()) { sum -= f.matchCost(); }
+        return sum;
+
+    case Data::type<OperatorAll>::value:
+        for (auto& f : operands()) { sum -= f.matchCost(); }
+        return sum;
+
+    case Data::type<OperatorNone>::value:
+        for (auto& f : operands()) { sum -= f.matchCost(); }
+        return sum;
+
+    case Data::type<Existence>::value:
+        // Equality and Range are more specific for increasing
+        // the chance to fail early check them before Existence
+        return 20;
+
+    case Data::type<Equality>::value:
+        return data.get<Equality>().global == FilterGlobal::undefined ? 10 : 1;
+
+    case Data::type<Filter::Range>::value:
+        return data.get<Range>().global == FilterGlobal::undefined ? 10 : 1;
+
+    case Data::type<Function>::value:
+        // Most expensive filter should be checked last
+        return 1000;
+    }
+    assert(false);
+    return 0;
+}
+
+const std::string& Filter::key() const {
+    static const std::string empty = "";
+
+    switch (data.get_type_index()) {
+
+    case Data::type<Existence>::value:
+        return data.get<Existence>().key;
+
+    case Data::type<Equality>::value:
+        return data.get<Equality>().key;
+
+    case Data::type<Filter::Range>::value:
+        return data.get<Range>().key;
+
+    default:
+        break;
+    }
+    return empty;
+}
+
+const std::vector<Filter>& Filter::operands() const {
+    static const std::vector<Filter> empty;
+
+    switch (data.get_type_index()) {
+    case Data::type<OperatorAny>::value:
+        return data.get<OperatorAny>().operands;
+
+    case Data::type<OperatorAll>::value:
+        return data.get<OperatorAll>().operands;
+
+    case Data::type<OperatorNone>::value:
+        return data.get<OperatorNone>().operands;
+
+    default:
+        break;
+    }
+    return empty;
+}
+
+int compareSetFilter(const Filter& a, const Filter& b) {
+    auto& oa = a.operands();
+    auto& ob = b.operands();
+
+    if (oa.size() != ob.size()) { return oa.size() < ob.size(); }
+
+    if (oa[0].data.is<Filter::Range>() &&
+        ob[0].data.is<Filter::Range>() &&
+        oa[0].key() == ob[0].key()) {
+        // take the one with more restrictive range
+        auto ra = oa[0].data.get<Filter::Range>();
+        auto rb = ob[0].data.get<Filter::Range>();
+
+        if (ra.max == std::numeric_limits<double>::infinity() &&
+            rb.max == std::numeric_limits<double>::infinity()) {
+
+            return rb.min - ra.min;
```

Comment:
It seems that this will prefer the more restrictive range in the case where neither range has a maximum, but in other cases there is no preference. Is this intentional? Feels like we could reasonable handle all cases.